File: WebClient SDK/src/default/abTest/abSettingsTileExtension.ts
import {
  TileExtension,
  ITileLocalExtensionContext,
  Tile,
  TileGroups,
  TileEvaluationEventArgs
} from 'tessa/ui/tiles';
import { userSession } from 'common/utility';
import {
  showError,
  showMessage,
  showConfirm,
  showLoadingOverlay,
  showNotEmpty,
  createDialogForm,
  createCardEditorModel
} from 'tessa/ui';
import { AdvancedCardDialogManager, CardToolbarAction } from 'tessa/ui/cards';
import { CardSection, systemKeyPrefix } from 'tessa/cards';
import { localize } from 'tessa/localization/localize';
import { CardRequest, CardService, CardResponse } from 'tessa/cards/service';
import { extension } from '@tessa/application';
import { FieldType, TypedField } from '@tessa/core';

@extension()
export class AbSettingsTileExtension extends TileExtension {
  public initializingLocal(context: ITileLocalExtensionContext): void {
    const panel = context.workspace.leftPanel;
    panel.tiles.push(
      new Tile({
        name: 'AbGenerateTestCards',
        caption: '$AbTest_Generator_GenerateTestCards',
        icon: 'ta icon-thin-001',
        contextSource: panel.contextSource,
        command: AbSettingsTileExtension.generateAction,
        group: TileGroups.CardsTop,
        order: 10,
        evaluating: AbSettingsTileExtension.enableOnSettingsCardAndAdministrator
      })
    );
  }

  private static enableOnSettingsCardAndAdministrator(e: TileEvaluationEventArgs) {
    const editor = e.current.context.cardEditor;

    e.setIsEnabledWithCollapsing(
      e.currentTile,
      !!editor &&
        !!editor.cardModel &&
        editor.cardModel.cardType.id === '35a03878-57b6-4263-ae36-92eb59032132' && // KrSettingsTypeID
        userSession.isAdmin
    );
  }

  private static async generateAction() {
    const dialogResult = await createDialogForm(
      'AbCardGenerator',
      undefined,
      undefined,
      undefined,
      async response => {
        const fields = response.card.sections.get('Table').fields;
        fields.set('UserCount', 0, FieldType.Int);
        fields.set('PartnerCount', 0, FieldType.Int);
      }
    );
    if (!dialogResult) {
      return;
    }

    const [form, model] = dialogResult;
    model.mainForm = form;
    const cardEditor = createCardEditorModel();
    await cardEditor.setCardModel(model);

    await AdvancedCardDialogManager.instance.showCard({
      editor: cardEditor,
      prepareEditorAction: editor => {
        editor.statusBarIsVisible = false;
        editor.toolbar.addItem(
          new CardToolbarAction({
            name: 'CreateCards',
            caption: '$AbTest_Generator_CreateCardsButton',
            icon: 'ta icon-Int426',
            command: async _ => {
              AbSettingsTileExtension.generateButtonAction(
                editor.cardModel!.card.sections.get('Table'),
                async () => await editor.close()
              );
            },
            order: 1
          })
        );

        editor.toolbar.addItem(
          new CardToolbarAction({
            name: 'Cancel',
            caption: '$UI_Common_Cancel',
            icon: 'ta icon-Int626',
            order: 2,
            toolTip: '$UI_Common_Cancel',
            command: async _ => {
              await editor.close();
            }
          })
        );

        editor.context.info[systemKeyPrefix + 'DialogClosingAction'] = () => Promise.resolve(false);
        return true;
      },
      dialogOptions: {
        withTabControlBackground: true
      },
      displayValue: model.card.typeCaption
    });
  }

  private static async generateButtonAction(section: CardSection, closeAction: () => void) {
    const fields = section.fields;
    const userCount = fields.get<number>('UserCount')!;
    if (userCount < 0) {
      await showError('$AbTest_Generator_WarnUserCountNegative');
      return false;
    }

    const partnerCount = fields.get<number>('PartnerCount')!;
    if (partnerCount < 0) {
      await showError('$AbTest_Generator_WarnPartnerCountNegative');
      return false;
    }

    if (userCount === 0 && partnerCount === 0) {
      await showMessage('$AbTest_Generator_WarnCardsCountNotDefined');
      return false;
    }

    const text =
      userCount === 0
        ? `${localize('$AbTest_Generator_CreatePartnersConfirmation')} ${partnerCount}. ${localize(
            '$UI_Common_ContinueConfirmation'
          )}`
        : partnerCount === 0
          ? `${localize('$AbTest_Generator_CreateUsersConfirmation')} ${userCount}. ${localize(
              '$UI_Common_ContinueConfirmation'
            )}`
          : `${localize('$AbTest_Generator_CreatePartnersConfirmation')} ${partnerCount}.\n` +
            `${localize('$AbTest_Generator_CreateUsersConfirmation')} ${userCount}.\n` +
            `${localize('$UI_Common_ContinueConfirmation')}`;

    if (!(await showConfirm(text))) {
      return false;
    }

    if (
      userCount > 1000 &&
      !(await showConfirm(
        `${localize('$AbTest_Generator_WarnTooMuchUsers')} ${userCount}. ${localize(
          '$UI_Common_ContinueConfirmation'
        )}`,
        '$UI_Common_Attention'
      ))
    ) {
      return false;
    }

    if (
      partnerCount > 1000 &&
      !(await showConfirm(
        `${localize('$AbTest_Generator_WarnTooMuchPartners')} ${partnerCount}. ${localize(
          '$UI_Common_ContinueConfirmation'
        )}`,
        '$UI_Common_Attention'
      ))
    ) {
      return false;
    }

    closeAction();

    let response!: CardResponse;
    await showLoadingOverlay(async () => {
      const request = new CardRequest();
      request.requestType = '207e75b5-abb8-403a-a12a-897019afccf6'; // AbTestData
      request.info = {
        UserCount: TypedField.createInt(userCount),
        PartnerCount: TypedField.createInt(partnerCount)
      };
      response = await CardService.instance.request(request);
    });

    const result = response.validationResult.build();
    await showNotEmpty(result);

    return true;
  }
}

File: Extensions/Tessa.Extensions.Default.Client/AbTest/AbSettingsTileExtension.cs
#nullable enable
using System;
using System.Threading;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared;
using Tessa.Extensions.Default.Shared.AbTest;
using Tessa.Platform;
using Tessa.Platform.Collections;
using Tessa.Platform.Runtime;
using Tessa.Platform.Storage;
using Tessa.Platform.Validation;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Tiles;
using Tessa.UI.Tiles.Extensions;

namespace Tessa.Extensions.Default.Client.AbTest
{
    /// <summary>
    /// Плитки для карточки с настройками типового решения.
    /// </summary>
    public sealed class AbSettingsTileExtension(
        CreateDialogFormFuncAsync createDialogFormFuncAsync,
        IAdvancedCardDialogManager cardDialogManager,
        ICardRepository cardRepository,
        ISession session)
        : TileExtension
    {
        #region Fields

        private readonly CreateDialogFormFuncAsync createDialogFormFuncAsync = NotNullOrThrow(createDialogFormFuncAsync);

        private readonly IAdvancedCardDialogManager cardDialogManager = NotNullOrThrow(cardDialogManager);

        private readonly ICardRepository cardRepository = NotNullOrThrow(cardRepository);

        private readonly ISession session = NotNullOrThrow(session);

        #endregion

        #region Private Methods

        private void EnableOnSettingsCardAndAdministrator(object? sender, TileEvaluationEventArgs e) =>
            e.SetIsEnabledWithCollapsing(
                e.CurrentTile,
                e.CurrentTile.Context.CardEditor?.CardModel?.CardType.ID == DefaultCardTypes.KrSettingsTypeID
                && this.session.User.IsAdministrator());

        private async Task<ValidationResult> GenerateAsync(int userCount, int partnerCount)
        {
            using (TessaSplash.Create(TessaSplashMessage.CreatingMultipleCards))
            {
                var response = await this.cardRepository.RequestAsync(
                    new CardRequest
                    {
                        RequestType = AbRequestTypes.TestData,
                        Info =
                        {
                            { "UserCount", Int32Boxes.Box(userCount) },
                            { "PartnerCount", Int32Boxes.Box(partnerCount) }
                        }
                    });

                return response.ValidationResult.Build();
            }
        }

        private async Task GenerateButtonFuncAsync(CardSection section, Func<Task> closeActionAsync)
        {
            var rawFields = section.RawFields;
            var userCount = rawFields.Get<int?>("UserCount") ?? 0;
            if (userCount < 0)
            {
                TessaDialog.ShowError("$AbTest_Generator_WarnUserCountNegative");
                return;
            }

            var partnerCount = rawFields.Get<int?>("PartnerCount") ?? 0;
            if (partnerCount < 0)
            {
                TessaDialog.ShowError("$AbTest_Generator_WarnPartnerCountNegative");
                return;
            }

            if (userCount == 0 && partnerCount == 0)
            {
                TessaDialog.ShowMessage("$AbTest_Generator_WarnCardsCountNotDefined");
                return;
            }

            var text = userCount == 0
                ? $"{await LocalizeNameAsync("AbTest_Generator_CreatePartnersConfirmation")} {partnerCount}. {await LocalizeNameAsync("UI_Common_ContinueConfirmation")}"
                : partnerCount == 0
                    ? $"{await LocalizeNameAsync("AbTest_Generator_CreateUsersConfirmation")} {userCount}. {await LocalizeNameAsync("UI_Common_ContinueConfirmation")}"
                    : $"{await LocalizeNameAsync("AbTest_Generator_CreatePartnersConfirmation")} {partnerCount}.{Environment.NewLine}{await LocalizeNameAsync("AbTest_Generator_CreateUsersConfirmation")} {userCount}.\r\n{await LocalizeNameAsync("UI_Common_ContinueConfirmation")}";

            if (!TessaDialog.Confirm(text))
            {
                return;
            }

            if (userCount > 1000
                && !TessaDialog.Confirm(
                    $"{await LocalizeNameAsync("AbTest_Generator_WarnTooMuchUsers")} {userCount}{await LocalizeNameAsync("UI_Common_ContinueConfirmation")}",
                    "$UI_Common_Attention"))
            {
                return;
            }

            if (partnerCount > 1000
                && !TessaDialog.Confirm(
                    $"{await LocalizeNameAsync("AbTest_Generator_WarnTooMuchPartners")} {partnerCount}{await LocalizeNameAsync("UI_Common_ContinueConfirmation")}",
                    "$UI_Common_Attention"))
            {
                return;
            }

            await closeActionAsync();

            var result = await this.GenerateAsync(userCount, partnerCount);
            TessaDialog.ShowNotEmpty(result);
        }

        #endregion

        #region Command Actions

        private async void GenerateActionAsync(object parameter)
        {
            var (_, model) = await this.createDialogFormFuncAsync("AbCardGenerator", modifyResponseAsync: (response, ct) =>
            {
                var fields = response.Card.Sections["Table"].RawFields;
                fields["UserCount"] = Int32Boxes.Zero;
                fields["PartnerCount"] = Int32Boxes.Zero;
                return ValueTask.CompletedTask;
            });
            if (model is null)
            {
                return;
            }

            await this.cardDialogManager.ShowCardAsync(
                model,
                prepareEditorActionAsync: (editor, ct) =>
                {
                    editor.StatusBarIsVisible = false;

                    editor.Toolbar.Actions.AddRange(
                        new CardToolbarAction(
                            "CreateCards",
                            "$AbTest_Generator_CreateCardsButton",
                            editor.Toolbar.CreateIcon("Int426"),
                            new DelegateCommand(async _ =>
                            {
                                await this.GenerateButtonFuncAsync(
                                    editor.CardModel.Card.Sections["Table"],
                                    closeActionAsync: async () => await editor.CloseAsync(cancellationToken: CancellationToken.None));
                            }),
                            order: 1),
                        new CardToolbarAction(
                            TileNames.Cancel,
                            "$UI_Common_Cancel",
                            editor.Toolbar.CreateIcon("Int626"),
                            new DelegateCommand(async _ => await editor.CloseAsync(cancellationToken: CancellationToken.None)),
                            tooltip: "$UI_Common_Cancel",
                            order: 2));

                    editor.Context.SetDialogClosingAction((dialogContext, args) => TaskBoxes.False);
                    return new(true);
                },
                options: new ShowCardOptions
                {
                    DisplayValue = model.Card.TypeCaption,
                    WithTabControlBackground = true
                });
        }

        #endregion

        #region Base Overrides

        public override Task InitializingGlobal(ITileGlobalExtensionContext context)
        {
            var panel = context.Workspace.LeftPanel;
            panel.Tiles.Add(
                new Tile(
                    "AbGenerateTestCards",
                    TileHelper.SplitCaption("$AbTest_Generator_GenerateTestCards"),
                    context.Icons.Get("Thin1"),
                    panel,
                    new DelegateCommand(this.GenerateActionAsync),
                    TileGroups.Top,
                    order: 1,
                    verticalAlignment: TileVerticalAlignment.Bottom,
                    evaluating: this.EnableOnSettingsCardAndAdministrator));

            return Task.CompletedTask;
        }

        #endregion
    }
}
